runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_i8mm.c

```diff
@@ -107,3 +107,249 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0(
 IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0(
     iree_uk_mmt4d_tile_s8s8s32_1x8x8_to_8x8x8_arm_64_i8mm,
     iree_uk_mmt4d_tile_s8s8s32_8x8x8_arm_64_i8mm, 8)
+
+IREE_UK_ATTRIBUTE_ALWAYS_INLINE inline void
+iree_uk_mmt4d_tile_s8s4s32_1x8x16_to_2x8x16_arm_64_i8mm(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 1 && M0 <= 2 && iree_uk_is_po2_u32(M0));
+  IREE_UK_ASSERT(!(params->K0 % 16));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+
+  const int8x16_t vmask = vmovq_n_s8(0xF0);
+  const int8x8_t vzero = vmov_n_s8(0);
+
+  int32x4_t acc[4];
+  IREE_UK_UNROLL for (int i = 0; i < 4; i++) {
+    // We start with zero accumulators and add the value of *out_ptr later.
+    // This is required for the int4 left shift described later.
+    acc[i] = vdupq_n_s32(0);
+  }
+
+  for (int k = 0; k < params->K; ++k) {
+    int8x16_t rhs[2][4];
+    IREE_UK_UNROLL for (int i = 0; i < 4; i++) {
+      int8x16_t r = vld1q_s8(rhs_ptr + 16 * i);
+      // We unpack int4s into individual int8s. To preserve signedness,
+      // int4s are moved to the upper 4-bits of each byte. This has the effect
+      // of multiplying each int4 by 2^4 = 16. To compensate, we divide the
+      // accumulator values by 16 before storing to memory.
+      // This int4 conversion trick is borrowed from the `qd8-f32-qc4w-gemm*`
+      // kernels in https://github.com/google/XNNPACK.
+      rhs[0][i] = vshlq_n_s8(r, 4);
+      rhs[1][i] = vandq_s8(r, vmask);
+    }
+    rhs_ptr += 64;
+
+    int8x16_t lhs[2];
+    if (M0 == 1) {
+      int8x8x2_t lhs_uzp = vld2_s8(lhs_ptr);
+      lhs_ptr += 16;
+      lhs[0] = vcombine_s8(lhs_uzp.val[0], vzero);
```

Comment:
this is a little slow.  if you made the 4 bit values on rhw planar you could ld2 lane s64 to transpose the lhs and not need a zip.
also your deleaved lhs require both shl and AND, but with planar you only need shl to start the first plane and can defer the AND.

Reply:
Agreed, the data layout could be more optimized. This is something we can revisit later since this would require non-trivial changes in the data-tiling logic.